lib/sink/connection/protocol.ex

```diff
@@ -2,47 +2,137 @@ defmodule Sink.Connection.Protocol do
   alias Sink.Connection.Protocol.Helpers
   alias Sink.Event
 
-  def encode_frame(:ack, message_id) do
-    message_type_id = 3
-    <<message_type_id::4, message_id::integer-size(12)>>
+  @typedoc "32-bit integer to uniquely identify a given server instance"
+  @type instance_id :: integer
+
+  @typedoc "Client application version in string format"
+  @type application_version :: String.t()
+
+  @typedoc "ID for a published message."
+  @type message_id :: pos_integer()
+
+  @type message ::
+          {:connection_request, {application_version, instance_id | nil}}
+          | {:connection_response,
+             :connected
+             | {:hello_new_client, instance_id}
+             | :instance_id_mismatch
+             | {:quarantined, payload :: binary}
+             | {:unsupported_protocol_version, term}
+             | :unsupported_application_version}
+          | {:ack, message_id}
+          | {:publish, message_id, payload :: binary}
+          | :ping
+          | :pong
+          | {:nack, message_id, payload :: binary}
+
+  @type nack_data() :: {binary(), String.t()}
+
+  # when we get to Sink 1.0 this will be 0 and we'll delete / deprecate anything
+  # 8 and higher
```

Comment:
@mdwaud why use `8` here? protocol version is encoded using 4 bytes, so `0..255`, so it feels a bit waistful to deprecate anything 8 or higher.